### PR TITLE
Prevent duplicate records in organizations list when creating a repository

### LINKED ISSUE #11258: organization listed multiple times when creating a repository
- Gitea version (or commit ref): 1.11.4
- Operating system: ubuntu 18.04
- Database (use `[x]`):
  - [x] PostgreSQL
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes - see e.g. https://try.gitea.io/org/testMaxOrg/teams 2 teams each with 'Create repository' permission

## Description
If an organization has multiple teams with 'Create repository' permission, then whenever a user belonging to several such teams tries to create a repository they see the organization name listed multiple times under "Owner" (see screenshot). The number of times matches the number of teams they are member of.
This is probably a bug, but at the same time it would be cool if it would show team names here, and the repository would be created and added to that team automatically...

## Screenshots
create

### PR DESCRIPTION
**Pull request by 6543**
prevent double entries in results of `GetOrgsCanCreateRepoByUserID`

I first try to only add GroupBy directly but xorm return broken user objects ...

... solution was to just query related UserIDs(OrgIDs) first and return OrgUsers based on this IDs

close #11258

### CODE REVIEW

**Comment by 6543** [comment]
I think I hit another xorm issue :(

`GetOrgsCanCreateRepoByUserID: mssql: Cannot call methods on nvarchar.`
`GetOrgsCanCreateRepoByUserID: pq: syntax error at or near "."`

File: models/org.go [hunk 1]
```diff
@@ -471,14 +471,20 @@ func GetOwnedOrgsByUserIDDesc(userID int64, desc string) ([]*User, error) {
...
 
-	return orgs, x.Join("INNER", "`team_user`", "`team_user`.org_id=`user`.id").
+	// because xorm do not return all cols, we have to collect IDs first and return user based on them afterwards
+	if err := x.Join("INNER", "`team_user`", "`team_user`.org_id=`user`.id").
 		Join("INNER", "`team`", "`team`.id=`team_user`.team_id").
 		Where("`team_user`.uid=?", userID).
 		And(builder.Eq{"`team`.authorize": AccessModeOwner}.Or(builder.Eq{"`team`.can_create_org_repo": true})).
-		Desc("`user`.updated_unix").
-		Find(&orgs)
+		Desc("`user`.updated_unix").GroupBy("user.id").Table("user").Cols("user.id").
```
Comment (6543) [inline models/org.go]:
```suggestion
		Desc("`user`.updated_unix").GroupBy("`user`.id").Table("`user`").Cols("`user`.id").
```

**Comment by zeripath** [comment]
How about:
```
// GetOrgsCanCreateRepoByUserID returns a list of organizations where given user ID
// are allowed to create repos.
func GetOrgsCanCreateRepoByUserID(userID int64) ([]*User, error) {
	orgs := make([]*User, 0, 10)

	return orgs, x.Where("id IN ("+
		"SELECT id FROM `user` "+
		"INNER JOIN `team_user` ON `team_user`.ord_id = `user`.id "+
		"INNER JOIN `team` on `team`.id = `team_user`.team_id "+
		"WHERE `team_user`.uid = ? AND (`team`.authorize = ? OR `team`.can_create_org_repo = ?) GROUP BY `user`.id)", userID, AccessModeOwner, true).
		Desc("`user`.updated_unix").
		Find(&orgs)
}
```

File: models/org.go [hunk 2]
```diff
@@ -471,14 +471,20 @@ func GetOwnedOrgsByUserIDDesc(userID int64, desc string) ([]*User, error) {
...
 		And(builder.Eq{"`team`.authorize": AccessModeOwner}.Or(builder.Eq{"`team`.can_create_org_repo": true})).
-		Desc("`user`.updated_unix").
-		Find(&orgs)
+		Desc("`user`.updated_unix").GroupBy("`user`.id").Table("`user`").Cols("`user`.id").
+		Find(&orgIDs); err != nil {
+		return nil, err
+	}
+
+	orgs := make([]*User, 0, len(orgIDs))
+	return orgs, x.In("id", orgIDs).Find(&orgs)
```
Comment (lafriks) [inline models/org.go]: This will break with too many bind arguments when user has many organizations

Comment (6543) [reply models/org.go]: @lafriks what about zeripaths version?

File: models/org.go [hunk 3]
```diff
@@ -471,20 +471,17 @@ func GetOwnedOrgsByUserIDDesc(userID int64, desc string) ([]*User, error) {
...
-		return nil, err
-	}
+	orgs := make([]*User, 0, 10)
 
-	orgs := make([]*User, 0, len(orgIDs))
-	return orgs, x.In("id", orgIDs).Find(&orgs)
+	return orgs, x.Where("id IN ("+
+		"SELECT `user`.id FROM `user` "+
+		"INNER JOIN `team_user` ON `team_user`.org_id = `user`.id "+
+		"INNER JOIN `team` on `team`.id = `team_user`.team_id "+
```
Comment (zeripath) [inline models/org.go]:
```suggestion
		"INNER JOIN `team` ON `team`.id = `team_user`.team_id "+
```

Comment (zeripath) [reply models/org.go]: It doesn't particularly matter but it's nice to have the SQL syntax in all capitals.

**Review by zeripath — APPROVED** [review]

File: models/org.go (hunk 2 above)
Comment (zeripath) [reply models/org.go]: TBH if it's likely that a person has more than 999 organizations that they can create in this UI is totally broken and we need to use a search box style instead

Comment (lafriks) [reply models/org.go]: I bet in many existing places we also thought same way when implementing them 😅

**Comment by lafriks** [comment]
I would prefer that builder would be used to build subselect to be  passed to in condition

**Comment by 6543** [comment]
@lafriks done

File: models/org.go [hunk 4]
```diff
@@ -473,15 +473,13 @@ func GetOwnedOrgsByUserIDDesc(userID int64, desc string) ([]*User, error) {
...
-		"(`team`.authorize = ? OR `team`.can_create_org_repo = ?) "+
-		"GROUP BY `user`.id)", userID, AccessModeOwner, true).
-		Desc("`user`.updated_unix").
-		Find(&orgs)
+	return orgs, x.Where(builder.In("id", builder.Select("`user`.id").From("`user`").
+		Join("INNER", "`team_user`", "`team_user`.org_id = `user`.id").
+		Join("INNER", "`team`", "`team`.id = `team_user`.team_id").
+		Where(builder.Eq{"`team_user`.uid": userID}).
+		And(builder.Eq{"`team`.authorize": AccessModeOwner}.Or(builder.Eq{"`team`.can_create_org_repo": true})).
+		GroupBy("`user`.id"))).
```
Comment (lafriks) [inline models/org.go]: When used in subselect there is no need for group by as most db engines will optimize it anyway to distinct select

Comment (guillep2k) [reply models/org.go]: In fact a `GROUP BY` can be slower in this case.

Comment (6543) [reply models/org.go]: ok ... will remove it

Comment (6543) [reply models/org.go]: done

**Comment by 6543** [comment]
something with the CI unrelated went wrong

**Review by zeripath — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Please send back port to v1.11

**Comment by 6543** [comment]
BP now :)

**Comment by 6543** [comment]
-> #11325

**Pull request merged**